Added documentation

sthirugn opened the pull request.

omaciel commented:
I have made some suggestions to @sthirugn in private and have asked him to incorporate some of it to this document. If you want to merge it as is and then add my suggestions, I'm fine with it too.

elyezer commented on .gitignore:
```diff
@@ -55,3 +55,7 @@ docs/_build/
 # PyBuilder
 target/
+
+# Misc
```
These can be placed outside the betelgeuse repo, since it is not related to the project is better to remove the ignore and use on the home for .pylarion and clone pylarion elsewhere.

sthirugn replied:
For me as a developer, pylarion is used only with betelgeuse and I like to keep betelgeuse and pylarion in one folder for my easy maintenance.  I thought this would help other developers too.  Having said that, I am happy to move this to my general .gitignore if this is not going to help fellow developers.

elyezer replied:
As you can we have two different uses, it is better to leave each developer add its own ignores outside.

elyezer commented on data_sample/tests/test_login_example.py:
```diff
@@ -0,0 +1,17 @@
+    def test_login_1(self):
+        """Check if a user is able to login with valid userid and password
+
+        @Feature: Login
```
Feature is not a token that betelgeuse looks, use requirement instead.

elyezer commented on data_sample/tests/test_login_example.py:
```diff
@@ -0,0 +1,17 @@
+# -*- encoding: utf-8 -*-
```
`-*- encoding: utf-8 -*-` is a format for emacs python default one is `coding=utf-8` better to use it.

For more information check https://www.python.org/dev/peps/pep-0263/.

elyezer commented on data_sample/tests/test_login_example.py:
```diff
@@ -0,0 +1,17 @@
+
+        @id: 60e48736-43a9-11e6-bcaa-104a7da122d7
+
+        @Assert: User is able to login successfully
```
Assert is not a token betelgeuse looks, use expectedresults instead.

elyezer commented on data_sample/tests/test_login_example.py:
```diff
@@ -0,0 +1,17 @@
+"""Test class for Login"""
+
+
+class LoginTestCase():
```
Since this is showing an example of a class which probably will be a unittest TestCase child is better to show it here. Otherwise a plain test function (test_loging_1) can be used.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+* Automatic creation/updation of Requirements and Test Cases from a Python
   project code base.
 * Creation of Test Runs based on a jUnit XML file.
```
This should also state that it needs the source code as well in order to proper map the ID.

sthirugn replied:
`Automatic creation/updation of Requirements and Test Cases from a Python project code base.` implies that this tool needs source code access.  The requirement creation mechanism is also detailed later in the doc which should cover this.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
 * Creation of Test Runs based on a jUnit XML file.
 
-Documentation contents:
+Betelgeuse uses `testimony <https://pypi.python.org/pypi/testimony>`_ and
```
Pylarion is not used to parse testcase. It is a client for the Polarion WSDL API. Will be more clear to change this to:

``` rst
Betelgeuse uses `testimony <https://pypi.python.org/pypi/testimony>`_ to parse test cases and Pylarion to talk with the Polarion's WSDL API.
```

sthirugn replied:
I have used the `respectively` construct to indicate that `testimony is used to parse test cases` and `pylarion is used to communicate with polarion`. 

Reference: https://www.quora.com/How-can-we-use-respectively-in-a-sentence-And-what-is-the-meaning-of-respectively-in-a-sentence

elyezer replied:
Never mind, I didn't saw the respectively on line 20. It was too late when I review this, I'm sorry.

sthirugn replied:
No worries at all.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+Login to polarion in a browser to check if the login user has permissions to do
+the following:
+
+* add/update test case
```
s/add/create let's normalize it and the line below.

sthirugn replied:
Sure I can update `create` instead of `add`.  What do you mean by `normalize` here?

elyezer replied:
I mean to normalize all terms to the same like using create and update everywhere.

sthirugn replied:
ok I think you mean `make it consistent`.  Please check the updated commit and let me know what do you think.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+* add/update test case
+* add/update requirements
+* create test runs
```
create/update test runs

Betelgeuse is able to update a previously created test run and will add just the results that is not already on the test run.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+       correctly with some versions of pylarion.  In this case, please use an
+       alternate working version of pylarion.
+     - Read pylarion documentation and set up .pylarion config file as
+       required.
```
A sample .pylarion file can be added here:

```
[webservice]
url=https://<polarion url>
svn_repo=https://<repo url>
user=<username>
password=<password>
default_project=<default_project>
```

Also worth mention that this file should be place at the user's home directory.

sthirugn replied:
I respectfully disagree. I like to keep betelgeuse project as atomic as possible.  This doc already says above that the user should read pylarion documentation to set .pylarion config file.  Any sample .pylarion file should live in the pylarion project README in my opinion.

Also pylarion documentation clearly says that the config file can live in current directory, user's home directory, or /etc/pylarion/pylarion.cfg.  I don't think we need to restrict betelgeuse users to keep the config in home directory only.

elyezer replied:
I suggested adding it here just as quick setup information, always consulting the official Pylarion documentation will be needed for more detailed information.

Feel free to add or not the the .pylarion sample here.

sthirugn replied:
Thank you for the response.  I think we can add it in future, as needed.  Having said this, we will have to improve Pylarion and testimony docs up-to-date too! This will help in easy reference of docs among the projects instead of duplicating the same info in multiple places.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+       $ cd betelgeuse
+       $ pip install -r requirements.txt
+       $ pip install -e .
```
requirements.txt already install Betelgeuse in editable mode. This line can be removed.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+   .. code-block:: bash
+
+       $ cd betelgeuse
```
Add a line above with `git clone https://github.com/SatelliteQE/betelgeuse.git`

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+.. note::
+
+  For easy understanding of betelgeuse, this repository is already included with
+  `data_sample` folder.  This folder contains sample tests and xml results which
```
better to name sample_project specially because both Betelgeuse and the test runner should be run on the same directory, this will ensure the test run ID mapping will work.

sthirugn replied:
Good point.  I can 
- leave the name as `data_sample`, if that is okay.
- add a note to say that the test runner and betelgeuse should be run on the same directory

elyezer replied:
I suggested to change the name because the word data is too generic and we are talking about a Python project, if we can mimic a real Python project with a clear name I think that would be better.

sthirugn replied:
I see your point and I think it will add value. It is updated now.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+test-case command
++++++++++++++++++
+Creates/Syncs test cases with polarion. This command performs the following
```
Create/update test cases on Polarion...

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+- Testimony is called to parse the test cases.
+- For each parsed test case, the following actions are performed:
+
+    - The requirement name is derived from the test module name.  If the test
```
It will first check if the test method have a `@Requirement` token and create/fetch the requirement after it. If that is not present then it will default to the module name.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+    - The requirement name is derived from the test module name.  If the test
+      module name is `test_login_example.py`, then the requirement name is
+      `Login Example`.
```
Inline code on RST is `content`.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+      module name is `test_login_example.py`, then the requirement name is
+      `Login Example`.
+    - Requirement is created.
+    - Test case is created and linked to the requirement.
```
This should be expanded saying about the default values and the expected tokens.

sthirugn replied:
Currently I am not much familiar about the betelgeuse project code, I can improve this once I start using betelgeuse more.

sthirugn replied:
This will be addressed in https://github.com/SatelliteQE/betelgeuse/issues/55

elyezer replied:
Fair enough. I can add that information later.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+.. code-block:: bash
+
+    $ betelgeuse test-case --path data_sample/tests/ PROJECT_CLOUD
```
I would just add the command call without the output, the output is easier to get outdated

sthirugn replied:
I have seen many open source projects in which the output was shown and it gave me confidence that I am doing the steps correctly.  I am following the same logic here to help any future betelgeuse learner.

Also I think it is the responsibility of the developer to keep this up to date, if we need to attract more new users.

elyezer replied:
Let's have this in mind when reviewing PRs. The review process for now on should strictly ensure updated documentation.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+.. code-block:: bash
+
+    $ betelgeuse test-run --path data_sample/results/sample-junit-result.xml \
```
In order to Betelgeuse proper map the `@ID` with the jUnit result it is needed that both the test runner and Betelgeuse get called on the same directory. With that said by having an entry like the following on the jUnit XML report:

``` xml
<testcase classname="data_sample.tests.test_login_example.LoginTestCase" file="data_sample/tests/test_login_example.py" line="421" name="test_login_1" time="694.768339396">...</testcase>
```

Betelgeuse will map the test method based on its classname and name by joining them with a dot: `data_sample.tests.test_login_example.LoginTestCase.test_login_1` then Betelgeuse will convert every test module path into a python import path for example `data_sample/tests/test_login_example.py` will become `data_sample.tests.test_login_example` and all test methods or functions will then be appended later on, for the method `test_login_1` inside the class `LoginTestCase` it will generate `data_sample.tests.test_login_example.LoginTestCase.test_login_1`. With the two information matching then it will check the `@ID` token of that test method or function and will query Polarion for the work item id that represents that ID, finally with the work_item_id information Betelgeuse will add the result for a Polarion test case (work item) into the test run.

It is important to note that if Betelgeuse can't find the `@ID` token information the ID will default to the Python import path, in the previous example it will be `data_sample.tests.test_login_example.LoginTestCase.test_login_1`.

If no result is returned when querying Polarion for a matching test case, that result will be skipped and the next entry on the jUnit will be processed. It is highly recommended to run test-case command before test-run to make sure all test cases are created/updated.

sthirugn replied:
This is a great explanation of the code logic.  I added this in docs now.

elyezer commented:
Please normalize all betelgeuse in lower case to Betelgeuse. The same can be done for Testimony and Polarion. Also jUnit XML is how Betelgeuse help message reports about the jUnit XML report.

elyezer commented on docs/index.rst:
```diff
@@ -1,14 +1,162 @@
+
+   .. note::
+
+     - It may be possible that the latest version of betelgeuse may not work
```
Currently using the alternate Pylarion version is a must since it adds ability to compare TestSteps which is needed in order to proper populate test steps for the test case.

sthirugn replied:
I have been told that some teams use original pylarion and other teams using forked version.  So this is a note to the user that he should be mindful of using the right pylarion version.  Also I am refraining from adding too much info on pylarion here as this is an internal project.

In the long run, hopefully our fixes get accepted in original pylarion and we don't need to fork it anymore.  So this is a kind of temporary solution.

elyezer replied:
Agreed, we can provide the detailed information on the right channels other than here.

sthirugn commented:
Thanks for all the valuable comments.  I will work on each of these and respond back.

elyezer commented:
@sthirugn thanks for the follow up and great documentation improvements. I have left some responses and observations about your points, really good points by the way.

sthirugn commented:
I believe I addressed all comments except for the open issues which will be addressed later.  Please let me know if I missed any comment.
- https://github.com/SatelliteQE/betelgeuse/issues/55
- https://github.com/SatelliteQE/betelgeuse/issues/56

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+Betelgeuse is a python program that reads standard Python test cases and offers
+tools to interact with Polarion. Possible interactions are:
+
+* Automatic creation/updation of Requirements and Test Cases from a Python
```
s/updation/update

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+* Automatic creation/updation of Requirements and Test Cases from a Python
   project code base.
-* Creation of Test Runs based on a jUnit XML file.
+* Automatic Creation/updation of Test Runs based on a jUnit XML file.
```
s/updation/update

elyezer replied:
s/Creation/creation

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+
+  2. Always run the test runner and Betelgeuse on the same directory to make
+  sure that the test run ID mapping works fine.  Otherwise Betelgeuse may
+  report ID errors.  More info in test_run_command_ section
```
More info can be found in...

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+      requirement name.  Otherwise it is derived from the test module name.  For
+      example, if the test module name is ``test_login_example``, then the
+      requirement name is ``Login Example``.
+    - Requirement is created.
```
First, we try to find the requirement by its name. If nothing is found, then we try to create one.

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+      example, if the test module name is ``test_login_example``, then the
+      requirement name is ``Login Example``.
+    - Requirement is created.
+    - Test case is created and linked to the requirement.
```
... or updated.

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+
+.. warning::
+
+   Are you not sure if you are using this command in a right way?  No
```
s/in a right way/correctly

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+file. This command performs the following steps:
+
+- Parses the jUnit XML file to read all the test cases and their run statuses.
+- Creates a new test run and updates it with all the parsed test case items and
```
iirc it will only create the test run if it cannot find one by name?

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+
+With the above report, Betelgeuse performs the following:
+
+- Derives the test method by joining its ``classname`` and ``name`` attributes
```
Derives the test method.. name? pseudo-id?

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+- Looks at the ``--source-code-path`` option value and converts every test
+  module path into a python import path.  For example,
+  ``sample_project/tests/test_login_example.py`` will become
+  ``sample_project.tests.test_login_example``.  All test methods or functions are
```
Should `All test methods...` be a separate bullet item on the same level?

omaciel commented on docs/index.rst:
```diff
@@ -1,14 +1,207 @@
+  then appended.  For example, the method ``test_login_1`` inside the class
+  ``LoginTestCase`` will be generated as
+  ``sample_project.tests.test_login_example.LoginTestCase.test_login_1``.
+- The information obtained from both the steps above are compared and ``@ID``
```
I find this part a bit confusing and hard to follow (for anyone who's not familiar with the source code or Betelgeuse). Can we file an issue to improve it a bit more?

sthirugn replied:
This is explaining the code logic in the same way as it is written.  Assumption is that this needs an in-depth understanding of how betelgeuse works.  Can you type some examples of how you think this can be improved?

omaciel commented:
@sthirugn thank you for updating this. I have added a few more nitpicks which I hope will make this first draft easier to understand and follow.

sthirugn commented:
Ready for review again.

omaciel commented:
ACK

elyezer commented:
@sthirugn thank you very much! This is awesome.
